vertx-mssql-client/src/main/java/io/vertx/mssqlclient/impl/codec/QueryCommandBaseCodec.java

```diff
@@ -18,53 +18,40 @@
 
 import java.util.stream.Collector;
 
-import static io.vertx.mssqlclient.impl.codec.EnvChange.*;
-import static io.vertx.mssqlclient.impl.utils.ByteBufUtils.readUnsignedByteLengthString;
-
 abstract class QueryCommandBaseCodec<T, C extends QueryCommandBase<T>> extends MSSQLCommandCodec<Boolean, C> {
+
+  protected int rowCount;
   protected RowResultDecoder<?, T> rowResultDecoder;
 
-  QueryCommandBaseCodec(C cmd) {
-    super(cmd);
+  QueryCommandBaseCodec(TdsMessageCodec tdsMessageCodec, C cmd) {
+    super(tdsMessageCodec, cmd);
   }
 
   private static <A, T> T emptyResult(Collector<Row, A, T> collector) {
     return collector.finisher().apply(collector.supplier().get());
   }
 
-  protected void encodeTransactionDescriptor(ByteBuf payload) {
-    payload.writeIntLE(18); // HeaderLength is always 18
-    payload.writeShortLE(0x0002); // HeaderType
-    payload.writeLongLE(encoder.transactionDescriptor);
-    payload.writeIntLE(1);
-  }
-
-  protected MSSQLRowDesc decodeColmetadataToken(ByteBuf payload) {
-    int columnCount = payload.readUnsignedShortLE();
-
-    ColumnData[] columnDatas = new ColumnData[columnCount];
-
-    for (int i = 0; i < columnCount; i++) {
-      long userType = payload.readUnsignedIntLE();
-      int flags = payload.readUnsignedShortLE();
-      DataType dataType = DataType.forId(payload.readUnsignedByte());
-      DataType.Metadata metadata = dataType.decodeMetadata(payload);
-      String columnName = readUnsignedByteLengthString(payload);
-      columnDatas[i] = new ColumnData(columnName, dataType, metadata);
-    }
-
-    return new MSSQLRowDesc(columnDatas);
+  @Override
+  protected void handleRowDesc(MSSQLRowDesc mssqlRowDesc) {
+    rowResultDecoder = new RowResultDecoder<>(cmd.collector(), mssqlRowDesc);
   }
 
+  @Override
   protected void handleRow(ByteBuf payload) {
     rowResultDecoder.handleRow(rowResultDecoder.desc.columnDatas.length, payload);
   }
 
+  @Override
   protected void handleNbcRow(ByteBuf payload) {
     rowResultDecoder.handleNbcRow(rowResultDecoder.desc.columnDatas.length, payload);
   }
 
-  protected void handleResultSetDone(int affectedRows) {
+  @Override
+  protected void handleAffectedRows(long count) {
+    rowCount += count;
```

Comment:
we could check `count >= 0` to catch programming error at this point

Reply:
There are a lot of other assertions we could add. As an example, when decoding parameters, we could check that scale/length are consistent with the `INT` type. I thought about it but couldn't see the value add.